File: 251004/Snopko/Publisher/src/main/java/com/example/restservice/controller/GlobalExceptionHandler.java
package com.example.restservice.controller;

import com.example.restservice.dto.response.ErrorResponseTo;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.util.NoSuchElementException;

@RestControllerAdvice
public class GlobalExceptionHandler {
    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ErrorResponseTo> handleValidationExceptions(MethodArgumentNotValidException ex) {
        ErrorResponseTo errorResponse = new ErrorResponseTo(ex.getMessage(), HttpStatus.BAD_REQUEST, 20);
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errorResponse);
    }

    @ExceptionHandler(DataIntegrityViolationException.class)
    public ResponseEntity<ErrorResponseTo> handleDataIntegrityViolationException(DataIntegrityViolationException ex) {
        ErrorResponseTo errorResponse = new ErrorResponseTo(ex.getMessage(), HttpStatus.FORBIDDEN, 30);
        return ResponseEntity.status(HttpStatus.FORBIDDEN).body(errorResponse);
    }

    @ExceptionHandler(NoSuchElementException.class)
    public ResponseEntity<ErrorResponseTo> handleNoSuchElementException(NoSuchElementException ex) {
        ErrorResponseTo errorResponse = new ErrorResponseTo(ex.getMessage(), HttpStatus.NOT_FOUND, 40);
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(errorResponse);
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorResponseTo> handleGeneralException(Exception ex) {
        ErrorResponseTo errorResponse = new ErrorResponseTo(ex.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR, 50);
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorResponse);
    }
}

File: 251003/Kukhatskavolets/microblog/publisher/src/main/kotlin/by/kukhatskavolets/publisher/controllers/GlobalExceptionHandler.kt
package by.kukhatskavolets.publisher.controllers

import by.kukhatskavolets.publisher.dto.responses.ExceptionResponseTo
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import java.sql.SQLException

@ControllerAdvice
class GlobalExceptionHandler {

    @ExceptionHandler(NoSuchElementException::class)
    fun handleNoSuchElementException(ex: NoSuchElementException): ResponseEntity<ExceptionResponseTo> {
        val httpStatus = HttpStatus.NOT_FOUND
        val errorCode = generateErrorCode(httpStatus, 1)
        val response = ExceptionResponseTo(errorCode, "Resource not found: ${ex.message}")
        return ResponseEntity(response, httpStatus)
    }

    @ExceptionHandler(SQLException::class)
    fun handlePSQLException(ex: SQLException): ResponseEntity<ExceptionResponseTo> {
        val httpStatus = HttpStatus.FORBIDDEN
        val errorCode = generateErrorCode(httpStatus, 2)
        val response = ExceptionResponseTo(errorCode, "Database access error: ${ex.message}")
        return ResponseEntity(response, httpStatus)
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationException(ex: MethodArgumentNotValidException): ResponseEntity<ExceptionResponseTo> {
        val httpStatus = HttpStatus.BAD_REQUEST
        val errorCode = generateErrorCode(httpStatus, 1) // 40001

        val errorMessage = ex.bindingResult.fieldErrors
            .joinToString("; ") { "${it.field}: ${it.defaultMessage}" }

        val response = ExceptionResponseTo(errorCode, "Validation failed: $errorMessage")
        return ResponseEntity(response, httpStatus)
    }

    private fun generateErrorCode(httpStatus: HttpStatus, logicalCode: Int): Int {
        return "${httpStatus.value()}%02d".format(logicalCode).toInt()
    }
}

